app/models/tag_set.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class TagSet < ApplicationRecord
+  belongs_to :tag_group, class_name: 'TagGroup', optional: false
+  belongs_to :tag2_group, class_name: 'TagGroup', optional: true
+
+  # We can assume adapter_type is the same for both tag groups
+  # But tag2_group may be present so we delegate to tag_group
+  delegate :adapter_type, to: :tag_group
+
+  validates :name, presence: true, uniqueness: true
+  validate :tag_group_adapter_types
```

Comment:
I think naming convention for validators is more like `validate :tag_group_adapter_types_must_match`

See examples here - https://apidock.com/rails/ActiveRecord/Validations/ClassMethods/validate